src/main/extension/extension_load.cpp

```diff
@@ -119,15 +152,66 @@ bool ExtensionHelper::TryInitialLoad(DBConfig &config, FileSystem &fs, const str
 		error = StringUtil::Format("Extension \"%s\" not found.\n%s", filename, message);
 		return false;
 	}
-	if (!config.options.allow_unsigned_extensions) {
-		auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
 
-		// signature is the last 256 bytes of the file
+	string metadata_segment;
+	metadata_segment.resize(512);
+
+	const std::string engine_version = std::string(GetVersionDirectoryName());
+	const std::string engine_platform = std::string(DuckDB::Platform());
+
+	auto handle = fs.OpenFile(filename, FileFlags::FILE_FLAGS_READ);
+
+	idx_t file_size = handle->GetFileSize();
+
+	if (file_size < 1024) {
+		throw InvalidInputException(
+		    "Extension \"%s\" do not have metadata compatible with DuckDB loading it "
+		    "(version %s, platform %s). File size in particular is lower than minimum threshold of 1024",
+		    filename, engine_version, engine_platform);
+	}
+
+	auto metadata_offset = file_size - metadata_segment.size();
+
+	handle->Read((void *)metadata_segment.data(), metadata_segment.size(), metadata_offset);
+
+	std::vector<std::string> metadata_field;
+	for (idx_t i = 0; i < 8; i++) {
+		metadata_field.emplace_back(metadata_segment, i * 32, 32);
+	}
 
-		string signature;
-		signature.resize(256);
+	std::reverse(metadata_field.begin(), metadata_field.end());
+
+	std::string extension_duckdb_platform = FilterZeroAtEnd(metadata_field[1]);
+	std::string extension_duckdb_version = FilterZeroAtEnd(metadata_field[2]);
+	std::string extension_version = FilterZeroAtEnd(metadata_field[3]);
+
+	string metadata_mismatch_error = "";
+	{
+		char a[32] = {0};
+		a[0] = '4';
+		if (strncmp(a, metadata_field[0].data(), 32) != 0) {
+			// metadata do not looks right, add this to the error message
+			metadata_mismatch_error =
+			    "\n" + StringUtil::Format("Extension \"%s\" do not have metadata compatible with DuckDB "
+			                              "loading it (version %s, platform %s)",
+			                              filename, engine_version, engine_platform);
+		} else if (engine_version != extension_duckdb_version || engine_platform != extension_duckdb_platform) {
+			metadata_mismatch_error = "\n" + StringUtil::Format("Extension \"%s\" (version %s, platfrom %s) does not "
```

Comment:
File \"%s\" is a DuckDB extension, but it was built for a different DuckDB version. The file was built for DuckDB version \"%s\", but we can only load extensions built for DuckDB version \"%s\".